Use only pytest

15r10nk opened the pull request:
I was able to execute all tests with pytest.
However assertion rewriting is still disabled for `test_main.py`.

But this has already several benefits:
* `tox -e py310 -- --sw` works
* no need to maintain two test environments

I also converted the `test_sample_files` to an parametrized test which makes it more useful for `pytest --sw`

I also don't know how I should handle the timeouts. 
I removed them for `test_sample_files` because they are not really useful for single file checks.
I kept them for `test_module_files` because this takes already a bit longer.
What's your opinion there?

I wanted to get your feedback before I continue.

alexmojaki commented on tests/test_main.py:
```diff
@@ -627,87 +632,89 @@ def is_annotation(node):
-    os.getenv('EXECUTING_SLOW_TESTS'),
-    'These tests are very slow, enable them explicitly',
-)
-class TestFiles(unittest.TestCase):
```
One advantage of the class is automatically applying the skip to both tests. See https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class

15r10nk replied:
ohh, I got a problem as long as the class derived from `unittest.TestCase`, but a normal class should work too.

alexmojaki commented on tests/test_main.py:
```diff
@@ -1359,23 +1366,25 @@ def foo():
 
 
-assert tester([1, 2, 3]) == [1, 2, 3]
+def test_tester():
```
The reason this stuff was global is that there was a time when global vs local mattered, and so this was there to test that things still worked globally. It's probably fine to move past that now, and the global stuff *was* irritating, but as we've seen with the recent decorator bug, weird things happen sometimes with bytecode.

15r10nk replied:
interesting, my problem was that i got a problem during test collection
```
================================================= ERRORS =================================================
__________________________________ ERROR collecting tests/test_main.py ___________________________________
tests/test_main.py:1370: in <module>
    assert tester([1, 2, 3]) == [1, 2, 3]
tests/utils.py:51: in __call__
    ex = self.get_executing(inspect.currentframe().f_back)
tests/utils.py:39: in get_executing
    return Source.executing(frame)
executing/executing.py:337: in executing
    node_finder = NodeFinder(frame, stmts, tree, lasti, source)
executing/executing.py:633: in __init__
    matching = list(self.matching_nodes(exprs))
executing/executing.py:699: in matching_nodes
    original_index = only(
executing/executing.py:167: in only
    raise NotOneValueFound('Expected one value, found 0')
E   executing.executing.NotOneValueFound: Expected one value, found 0
======================================== short test summary info =========================================
ERROR tests/test_main.py - executing.executing.NotOneValueFound: Expected one value, found 0
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

Does it look familiar to you?
I look for the reason and try to make it work.

alexmojaki replied:
Sounds like either:

1. pytest is still rewriting assertions
2. assertions have problems even without pytest rewriting
3. something at the global level really does make a difference still

15r10nk replied:
I have `test_main.py` minimized to the following

```
# -*- coding: utf-8 -*-
"""

assert rewriting will break executing
PYTEST_DONT_REWRITE
"""
import pytest


from tests.utils import tester


import inspect
frame=inspect.currentframe()
import dis
dis.dis(frame.f_code)

assert tester([1, 2, 3]) == [1, 2, 3]

```

Which fails with the exception from above.
The assert rewriting is correct disabled.
The thing is that removing `import pytest` makes it work.
`pytest` is not used anywhere and the generated bytecode looks identical except the missing import.

I will stop here. Maybe I have tomorrow an idea ... or you.

alexmojaki replied:
Ah, that makes sense.

https://github.com/alexmojaki/executing/blob/11648389b9c8c3b59ff1137f64d1fc7c0651cb40/executing/executing.py#L543-L550

Seems like we would need a better way to detect (for the sake of avoiding/ignoring) pytest assertion rewriting.

15r10nk replied:
I check the bytecode for LOAD ops with "@py_assert" as variable name. What do you think?

alexmojaki reviewed (changes requested):
Thanks, this looks very cool!

alexmojaki commented:
> I also don't know how I should handle the timeouts.

I think they're not necessary any more since we don't use travis.

alexmojaki commented on tests/test_pytest.py:
```diff
@@ -399,3 +399,21 @@ def collect_names(code):
+import dis
+
+def test_pytest_rewrite():
+    frame=inspect.currentframe()
+    #dis.dis(frame.f_code)
+    for bc in get_instructions(frame.f_code):
+        print(bc)
+    
+    # check for assert statements rewrite caused by this assert
+    assert is_rewritten_by_pytest(frame.f_code)==True
```
```suggestion
    frame = inspect.currentframe()

    # check for assert statements rewrite caused by this assert
    assert is_rewritten_by_pytest(frame.f_code)
```

alexmojaki commented on tests/test_pytest.py:
```diff
@@ -399,3 +399,21 @@ def collect_names(code):
+    frame=inspect.currentframe()
+
+    # no assert -> no rewrite
+    if is_rewritten_by_pytest(frame.f_code):
+        raise AssertionError("unexpected pytest assert rewrite")
```
```suggestion
    assert not is_rewritten_by_pytest(frame.f_code)
```

15r10nk replied:
This will not work. Pytest will rewrite the assert and the test will fail. 

The new check checks only for rewritten assertions in the function and not in the module. Is this ok for the SentinelParser?

alexmojaki replied:
> This will not work. Pytest will rewrite the assert and the test will fail.

lol good point

> The new check checks only for rewritten assertions in the function and not in the module. Is this ok for the SentinelParser?

Yes, that should be OK, and that should be the same as the previous behavious since that also only looked at the current frame's code object.

alexmojaki commented on tests/test_main.py:
```diff
@@ -1359,23 +1362,9 @@ def foo():
-assert (tester or 234) == 234
-assert (tester and 1123) is tester
+def test_tester():
+    # tester calls should be tested at global scope
```
Please also rename the test to `test_global_tester_calls`

alexmojaki commented on tests/test_main.py:
```diff
@@ -1247,7 +1250,7 @@ def check_code(self, code, nodes, decorators, check_names):
             # The relation between `ast.Name` and `argval` is already
             # covered by the verifier and much more complex in python 3.11 
```
A bit off topic, but mangling of variable names has always been there and I just didn't realise it, right? It didn't become more complex in 3.11?

15r10nk replied:
I learned a lot of new stuff during this pull-request, but I don't know how it has been before. Name mangeling also surprised me

15r10nk replied:
I experimented a bit:

I can not think of any use for mangled local variables in real code.

So I tried if it is possible to use mangled arguments and the result surprised me.

script:
```python
class A:
    def a(self, *, __private): # argument is mangled
        print(__private)

    def b(self, v):
        self.a(__private=v) # call is not mangled

    def c(self, v):
        self.a(_A__private=v) # we have to call it explicit


a = A()

a.c("c")
a.a(_A__private="a")
a.b("b")

```

output (Python 3.11.0rc2+):
```python
c
a
Traceback (most recent call last):
  File "/home/frank/projects/executing/except_test.py", line 16, in <module>
    a.b("b")
  File "/home/frank/projects/executing/except_test.py", line 6, in b
    self.a(__private=v) # call is not mangled
    ^^^^^^^^^^^^^^^^^^^
TypeError: A.a() got an unexpected keyword argument '__private'
```
I don't think that this is intentional, because it is not possible to call this function with the `__private` argument.

alexmojaki replied:
This is an excellent point. It already doesn't feel intentional just because it's weird, but this is a solid example of a bad outcome, `self.a(__private=v)` not working can only be fairly described as a bug and should be reported as such. On the other hand, it seems very low priority, because no one does this in practice, to the extent that we didn't know about it. I wouldn't be surprised if there's already a report about this and no one really cares. But I might be wrong.

15r10nk replied:
found an interresting real world example:
``` python
class Opt:
    __optionalNotMatched = _NullToken()

    def __init__(
        self, expr: Union[ParserElement, str], default: Any = __optionalNotMatched
    ):
        ...
```
there are also several other

``` bash
$ rg " __[A-Za-z_]*[^_,:] " **.py
```

I don't think that it is worth thinking to much about that stuff.

ok ... a last one :-) 

``` python
# name mangling
class TestMangling:

    def test(self):
        try:
            raise Exception(10)
        except Exception as __e:
            assert __e == __e

```

alexmojaki commented on executing/executing.py:
```diff
@@ -534,17 +534,20 @@ def compile_similar_to(source, matching_code):
 
+def is_rewritten_by_pytest(code):
+    return any(
+        bc.opname.startswith("LOAD_") and isinstance(bc.argval,str) and bc.argval.startswith("@py_assert")
```
```suggestion
        isinstance(bc.argval,str) and bc.argval.startswith("@py")
```

Does this also work?

alexmojaki commented on .github/workflows/test.yml:
```diff
@@ -25,8 +25,7 @@ jobs:
       run: |
-        coverage run --include=executing/executing.py -m unittest tests.test_main
-        coverage run --include=executing/executing.py --append -m pytest tests/test_pytest.py
+        coverage run --include=executing/executing.py --append -m pytest tests
```
At some point this needs to switch to including the whole executing folder, while accounting for the version stuff in `__init__`

15r10nk replied:
I don't understand what you mean here? 
There are no tests in the executing folder.

alexmojaki replied:
Sorry, I see how it wasn't clear what I was talking about, and in any case it's not relevant to this PR. See #52.

15r10nk commented:
I rebased it and checked for LOAD_CONST (... see unit tests)

should I exclude _pytest stuff from the test_module_files? The tests failed once for some pytest source file.

alexmojaki commented on tests/test_pytest.py:
```diff
@@ -399,3 +399,28 @@ def collect_names(code):
         ) == {"Test","_","a", "self", "__thing"}
 
 
+from executing.executing import is_rewritten_by_pytest,get_instructions
+import dis
```
Please put these at the top

alexmojaki commented:
> should I exclude _pytest stuff from the test_module_files? The tests failed once for some pytest source file.

That sounds like a bug to investigate.

Besides, I see a new failure after that.

Also, the previous build failed to report the incorrect node: `AttributeError: 'Subscript' object has no attribute 'end_lineno'`

15r10nk commented:
I minimized the failure down to the following. (currently working on the issue minimization)

``` python
class Block(ABC):
    def __init__(self, container: Optional["Container"] = None):
        self._lines: List[str] = []

```

The annotated assignment seems to be the problem.
Can you look into it, because it is part of the 3.11 code.

alexmojaki commented:
I think all you need to do is check `is_annotation` starting from 3.6 (when the syntax was introduced) rather than 3.8.

15r10nk commented:
I added looked at the failure and the problem is that the pre 3.11 test code does not analyse deadcode.

example:
``` python
class A:
    def __cmp__(self, other):
        if type(other) is self.__class__:
            if self is other:
                return 0
            return -1
        return NotImplemented
        raise TypeError("unorderable types: %s() and %s()" % (self.__class__.__name__, other.__class__.__name__))
```
There is no bytecode for the TypeError ast-node.

I added the test also to the samples for now but I think I will remove it later again.

Should I continue here and fix this problems?
I think it might be better to skip all files which create problems and fix them in another issue.
Maybe even skip the module_tests for python < 3.11, because there could be much more of this kind of errors and the module_tests are not so deterministic.

alexmojaki commented:
How hard do you think it'd be to get deadcode analysis working for older versions?

15r10nk commented:
I don't know. The problem is that different python implementations could decide differently what is deadcode. 
However, I think it and is necessary for #42, because I would like to test it with a similar file count like #31.

alexmojaki commented:
> I don't know. The problem is that different python implementations could decide differently what is deadcode.

Assuming 3.11 is the most optimised, I think it'd be a good compromise to always ignore nodes marked as deadcode when they don't match bytecode, but only complain about bytecode matching deadcode in 3.11. That might make things easier.

15r10nk commented:
Therefore I would have to make the deadcode analysis work down to python 2.7 ...

I would like to improve the tests before I start to change there anything again (see #55).

I try to ignore the enum module for now, if this is also ok.

alexmojaki commented on tests/test_main.py:
```diff
@@ -708,6 +711,9 @@ def test_module_files(self):
                     # The sentinel actually appearing in code messes things up
                     'executing' in filename
+                    # ignore pytest files because they break the SentinelNodeFinder
+                    or '_pytest' in filename
```
I think this file stopped being a problem once annotations were considered in 3.6+

15r10nk replied:
no problem any more.

alexmojaki replied:
So these lines should be removed, right?

15r10nk replied:
sorry you are right. I forgot that.

alexmojaki replied:
Ah, it just hit a genuine failure because it doesn't know how to distinguish between the comprehension codes in [`{t[0] for t in lines2} - {t[0] for t in lines1}`](https://github.com/pytest-dev/pytest/blob/2be1b8f3559570c456f4bf64fea8067e368dcdfd/src/_pytest/pytester.py#L172). This is covered in the normal tests. So `pytester.py` specifically should be excluded.

alexmojaki replied:
sigh, another pypy optimisation. can you skip if `PYPY and '__debug__' in source.text'`?

if it fails again you can go back to skipping all pytest files

alexmojaki commented on tests/test_main.py:
```diff
@@ -708,6 +711,9 @@ def test_module_files(self):
                     'executing' in filename
+                    # ignore pytest files because they break the SentinelNodeFinder
+                    or '_pytest' in filename
+                    or 'enum' in filename
```
This file seems to only be a problem in pypy, as they seem to have had deadcode analysis for much longer than CPython. Can you move this case inside `or PYPY and ('sysconfig.py' in filename or 'pyparsing.py' in filename)` just below?

15r10nk replied:
I changed it.

15r10nk commented:
this expression gets evaluated at compile time ... `__debug__` is known and the literals get evaluated ...
``` python
PYC_EXT = ".py" + (__debug__ and "c" or "o")
```

alexmojaki commented:
Thanks!
